Dockerfile

```diff
@@ -1,25 +1,25 @@
 # First step is to generate a static build for the client
-FROM node:12-alpine AS BUILD_CLIENT
+FROM node:14 AS BUILD_CLIENT
```

Comment:
could we use 14-alpine to get smaller images?

Reply:
I've tried `node:16-alpine` for sure, as you can see in following screenshots,   
however it resulted in a "missing `query-engine` binary" error on ARM.


<details>
<summary markdown="block">
Show/Hide `node:16-alpine` build log screenshots
</summary>

![node-16-alpine-build-log-1](https://user-images.githubusercontent.com/24852597/188505250-136c4949-6c67-4e71-9af5-f1d33878807f.png)

![node-16-alpine-build-log-2](https://user-images.githubusercontent.com/24852597/188505259-57d7c6a4-9d20-4de4-9770-d061b5ab1711.png)


</details>

I think I also tried `node:14-alpine` with the same issue.   
With the full `node:14` base image this error went away.

Reply:
the size of the base node image is really huge, could you try to use the slim version as suggested by this issue ? https://github.com/prisma/prisma/issues/8478

Reply:
Sadly `node:14-slim` didn't work either... :disappointed:    
It results in following error: `Error: Unknown binaryTarget linux-arm64-openssl-undefined and no custom binaries were provided`


<details>
<summary markdown="block">
Show/Hide full `node:14-slim` build log
</summary>

```log
git push ledokku development
============================

Enumerating objects: 5, done.
Counting objects: 100% (5/5), done.
Delta compression using up to 12 threads
Compressing objects: 100% (3/3), done.
Writing objects: 100% (3/3), 345 bytes | 345.00 KiB/s, done.
Total 3 (delta 2), reused 0 (delta 0), pack-reused 0
-----> Cleaning up...
-----> Building ledokku from Dockerfile
-----> Setting config vars
       DOKKU_DOCKERFILE_PORTS:  4000
remote: build context to Docker daemon  7.895MB
Step 1/34 : FROM node:14-slim AS BUILD_CLIENT
 ---> 6ef5be5eda09
Step 2/34 : WORKDIR /usr/src/app
 ---> Running in 1db49fa7dce1
Removing intermediate container 1db49fa7dce1
 ---> 3bfa44c6c319
Step 3/34 : COPY package.json yarn.lock ./
 ---> 792456ca9416
Step 4/34 : COPY client/package.json ./client/package.json
 ---> f64b1fbda821
Step 5/34 : RUN yarn install --frozen-lockfile
 ---> Running in d10be51cef3f
yarn install v1.22.19
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
[4/5] Linking dependencies...
[5/5] Building fresh packages...
Done in 192.42s.
Removing intermediate container d10be51cef3f
 ---> 2f7936dec460
Step 6/34 : COPY client ./client
 ---> d319ba5f7bd5
Step 7/34 : RUN cd client; yarn build
 ---> Running in 0ebcf6b873fb
yarn run v1.22.19
$ yarn build:css && SKIP_PREFLIGHT_CHECK=true react-scripts build
$ postcss src/styles/index.css -o src/generated/index.css
Creating an optimized production build...
Compiled successfully.

File sizes after gzip:

  294.1 KB   build/static/css/main.a608e9fb.chunk.css
  285.52 KB  build/static/js/2.fe905f83.chunk.js
  32.75 KB   build/static/js/main.c93ce56a.chunk.js
  783 B      build/static/js/runtime-main.757a9811.js

The project was built assuming it is hosted at /.
You can control this with the homepage field in your package.json.

The build folder is ready to be deployed.
You may serve it with a static server:

  npm install -g serve
  serve -s build

Find out more about deployment here:

  https://cra.link/deployment

Done in 260.38s.
Removing intermediate container 0ebcf6b873fb
 ---> 656f5b0e0092
Step 8/34 : FROM node:14-slim AS BUILD_SERVER
 ---> 6ef5be5eda09
Step 9/34 : RUN apt update; apt install -y openssl
 ---> Running in 17821f069da6

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Get:1 http://deb.debian.org/debian buster InRelease [122 kB]
Get:2 http://deb.debian.org/debian-security buster/updates InRelease [34.8 kB]
Get:3 http://deb.debian.org/debian buster-updates InRelease [56.6 kB]
Get:4 http://deb.debian.org/debian buster/main arm64 Packages [7739 kB]
Get:5 http://deb.debian.org/debian-security buster/updates/main arm64 Packages [341 kB]
Get:6 http://deb.debian.org/debian buster-updates/main arm64 Packages [8780 B]
Fetched 8303 kB in 4s (2310 kB/s)
Reading package lists...
Building dependency tree...
Reading state information...
2 packages can be upgraded. Run 'apt list --upgradable' to see them.

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
The following package was automatically installed and is no longer required:
  lsb-base
Use 'apt autoremove' to remove it.
The following additional packages will be installed:
  libssl1.1
Suggested packages:
  ca-certificates
The following NEW packages will be installed:
  libssl1.1 openssl
0 upgraded, 2 newly installed, 0 to remove and 2 not upgraded.
Need to get 2225 kB of archives.
After this operation, 5282 kB of additional disk space will be used.
Get:1 http://deb.debian.org/debian-security buster/updates/main arm64 libssl1.1 arm64 1.1.1n-0+deb10u3 [1392 kB]
Get:2 http://deb.debian.org/debian-security buster/updates/main arm64 openssl arm64 1.1.1n-0+deb10u3 [834 kB]
debconf: delaying package configuration, since apt-utils is not installed
Fetched 2225 kB in 0s (5767 kB/s)
remote: g previously unselected package libssl1.1:arm64.
remote: (Reading database ... 6462 files and directories currently installed.)
remote: g to unpack .../libssl1.1_1.1.1n-0+deb10u3_arm64.deb ...
remote: g libssl1.1:arm64 (1.1.1n-0+deb10u3) ...
remote: g previously unselected package openssl.
remote: g to unpack .../openssl_1.1.1n-0+deb10u3_arm64.deb ...
remote: g openssl (1.1.1n-0+deb10u3) ...
remote: up libssl1.1:arm64 (1.1.1n-0+deb10u3) ...
remote:  unable to initialize frontend: Dialog
remote:  (TERM is not set, so the dialog frontend is not usable.)
remote:  falling back to frontend: Readline
remote:  unable to initialize frontend: Readline
debconf: (Can't locate Term/ReadLine.pm in @INC (you may need to install the Term::ReadLine module) (@INC contains: /etc/perl /usr/local/lib/aarch64-linux-gnu/perl/5.28.1 /usr/local/share/perl/5.28.1 /usr/lib/aarch64-linux-gnu/perl5/5.28 /usr/share/perl5 /usr/lib/aarch64-linux-gnu/perl/5.28 /usr/share/perl/5.28 /usr/local/lib/site_perl /usr/lib/aarch64-linux-gnu/perl-base) at /usr/share/perl5/Debconf/Fremote: Readline.pm line 7.)
remote:  falling back to frontend: Teletype
remote: up openssl (1.1.1n-0+deb10u3) ...
remote: ng triggers for libc-bin (2.28-10+deb10u1) ...
Removing intermediate container 17821f069da6
 ---> 1a4edb2a3270
Step 10/34 : WORKDIR /usr/src/app
 ---> Running in 2273b1ab3f88
Removing intermediate container 2273b1ab3f88
 ---> 0f5bab9ac25b
Step 11/34 : COPY package.json yarn.lock ./
 ---> a795c7a6d978
Step 12/34 : COPY server/package.json ./server/package.json
 ---> d53725e81f72
Step 13/34 : COPY server/prisma ./server/prisma
 ---> 4e02a9a03b9d
Step 14/34 : RUN yarn install --frozen-lockfile
 ---> Running in 5f283b8e7501
yarn install v1.22.19
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
[4/5] Linking dependencies...
[5/5] Building fresh packages...
Done in 138.32s.
Removing intermediate container 5f283b8e7501
 ---> 2ba75232b734
Step 15/34 : COPY server ./server
 ---> 780af40c2b3b
Step 16/34 : RUN cd server; yarn build
 ---> Running in 0dea056dc207
yarn run v1.22.19
$ tsc
Done in 22.52s.
Removing intermediate container 0dea056dc207
 ---> c2ab08ee599f
Step 17/34 : FROM node:14-slim
 ---> 6ef5be5eda09
Step 18/34 : WORKDIR /usr/src/app
 ---> Using cache
 ---> 3bfa44c6c319
Step 19/34 : COPY package.json yarn.lock ./
 ---> Using cache
 ---> 792456ca9416
Step 20/34 : COPY server/package.json ./server/package.json
 ---> 167e70f1970c
Step 21/34 : COPY server/prisma ./server/prisma
 ---> 379917ed6baa
Step 22/34 : COPY --from=BUILD_CLIENT /usr/src/app/client/build ./client/build
 ---> 087acc4ca7ab
Step 23/34 : COPY --from=BUILD_SERVER /usr/src/app/server/build ./server/build
 ---> 8d8126ade2d8
Step 24/34 : ENV NODE_ENV production
 ---> Running in ea35ce8000bd
Removing intermediate container ea35ce8000bd
 ---> 043e31b30d0e
Step 25/34 : RUN yarn install --frozen-lockfile --production
 ---> Running in d9018a6eb22d
yarn install v1.22.19
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
[4/5] Linking dependencies...
[5/5] Building fresh packages...
Done in 69.64s.
Removing intermediate container d9018a6eb22d
 ---> 0faba018609a
Step 26/34 : WORKDIR /usr/src/app/server
 ---> Running in 8e8fb98bb9d8
Removing intermediate container 8e8fb98bb9d8
 ---> cb062d4c8a20
Step 27/34 : EXPOSE 4000
 ---> Running in 667ea2ee4e33
Removing intermediate container 667ea2ee4e33
 ---> ceef0aa0fff6
Step 28/34 : CMD [ "yarn", "start" ]
 ---> Running in 44a688a848ce
Removing intermediate container 44a688a848ce
 ---> 8071f13da6c2
Step 29/34 : LABEL com.dokku.app-name=ledokku
 ---> Running in 841b7209da6d
Removing intermediate container 841b7209da6d
 ---> 12e58e348dc9
Step 30/34 : LABEL com.dokku.builder-type=dockerfile
 ---> Running in 85d84b1953fc
Removing intermediate container 85d84b1953fc
 ---> e4a3d978f199
Step 31/34 : LABEL com.dokku.image-stage=build
 ---> Running in 4de7bf638c1b
Removing intermediate container 4de7bf638c1b
 ---> 5aed74a3970f
Step 32/34 : LABEL dokku=
 ---> Running in b54b12f05fd1
Removing intermediate container b54b12f05fd1
 ---> ef6ccde83599
Step 33/34 : LABEL org.label-schema.schema-version=1.0
 ---> Running in dee7c16108e4
Removing intermediate container dee7c16108e4
 ---> 45506f90a13d
Step 34/34 : LABEL org.label-schema.vendor=dokku
 ---> Running in 2a9cb4bfdd65
Removing intermediate container 2a9cb4bfdd65
 ---> fbac38927081
Successfully built fbac38927081
Successfully tagged dokku/ledokku:latest
-----> Releasing ledokku...
-----> Checking for predeploy task
       No predeploy task found, skipping
-----> Checking for release task
       No release task found, skipping
-----> No Procfile found in app image
=====> Processing deployment checks
       No CHECKS file found. Simple container checks will be performed.
       For more efficient zero downtime deployments, create a CHECKS file. See https://dokku.com/docs/deployment/zero-downtime-deploys/ for examples
-----> Deploying ledokku via the docker-local scheduler...
-----> Deploying web (count=1)
       Attempting pre-flight checks (web.1)
       Waiting for 10 seconds (web.1)
a3be8da4ae1885b5360924e1898c5c9e3d396bfa036062a95c3439326563a6e3
remote:  !     App container failed to start (web.1)
=====> Start of ledokku container output (web.1)
       yarn run v1.22.19
       $ yarn prisma:migrate && node build/index.js
       $ prisma migrate deploy --preview-feature
       Error: Unknown binaryTarget linux-arm64-openssl-undefined and no custom binaries were provided
       error Command failed with exit code 1.
       info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
       error Command failed with exit code 1.
       info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
       yarn run v1.22.19
       $ yarn prisma:migrate && node build/index.js
       $ prisma migrate deploy --preview-feature
       Error: Unknown binaryTarget linux-arm64-openssl-undefined and no custom binaries were provided
       error Command failed with exit code 1.
       info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
       error Command failed with exit code 1.
       info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
       yarn run v1.22.19
       $ yarn prisma:migrate && node build/index.js
       $ prisma migrate deploy --preview-feature
       Error: Unknown binaryTarget linux-arm64-openssl-undefined and no custom binaries were provided
       error Command failed with exit code 1.
       info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
       error Command failed with exit code 1.
       info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
       yarn run v1.22.19
       $ yarn prisma:migrate && node build/index.js
=====> End of ledokku container output (web.1)
remote: parallel: This job failed:
remote: /var/lib/dokku/plugins/available/scheduler-docker-local/bin/scheduler-deploy-process-container ledokku dockerfile dokku/ledokku:latest latest web 1 1
To my-website.com:ledokku
 ! [remote rejected] development -> development (pre-receive hook declined)
error: failed to push some refs to 'my-website.com:ledokku'
```

</details>

I recall trying a lot of different things, and initially didn't want to use the full node image either.   
However it was the only way I've managed to get things working out on ARM.

Reply:
And same with node 16 or 18 (slime or alpine version)? I would really like to avoid using the raw node image if possible